src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -0,0 +1,574 @@
+      isEqualType(tp1, tp2) && ss1.zip(ss2).forall((subspace _).tupled)
+    case (Const(v1, _), Const(v2, _)) => v1 == v2
+    case (Const(_, tp1), Typ(tp2, _)) => isSubType(tp1, tp2)
+    case (Const(_, _), Or(ss)) => ss.exists(subspace(a, _))
+    case (Const(_, _), _) => false
+    case (_, Const(_, _)) => false
+    case (Var(x, _), Var(y, _)) => x == y
+    case (Var(_, tp1), Typ(tp2, _)) => isSubType(tp1, tp2)
+    case (Var(_, _), Or(ss)) => ss.exists(subspace(a, _))
+    case (Var(_, _), _) => false
+    case (_, Var(_, _)) => false
+  }
+
+  /** Intersection of two spaces  */
+  def intersect(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) | (_, Empty) => Empty
+    case (_, Or(ss)) => Or(ss.map(intersect(a, _)))
+    case (Or(ss), _) => Or(ss.map(intersect(_, b)))
+    case (Typ(tp1, _), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) a
+      else if (isSubType(tp2, tp1)) b
+      else Empty
+    case (Typ(tp1, _), Kon(tp2, ss)) =>
+      if (isSubType(tp2, tp1)) b
+      else if (isSubType(tp1, tp2)) a
+      else Empty
+    case (Kon(tp1, ss), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2) || isSubType(tp2, tp1)) a
+      else Empty
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      if (!isEqualType(tp1, tp2)) Empty
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) Empty
+      else
+        Kon(tp1, ss1.zip(ss2).map((intersect _).tupled))
+    case (Const(v1, _), Const(v2, _)) =>
+      if (v1 == v2) a else Empty
+    case (Const(_, tp1), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) a else Empty
+    case (Const(_, _), _) => Empty
+    case (Typ(tp1, _), Const(_, tp2)) =>
+      if (isSubType(tp2, tp1)) b else Empty
+    case (_, Const(_, _)) => Empty
+    case (Var(x, _), Var(y, _)) =>
+      if (x == y) a else Empty
+    case (Var(_, tp1), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) a else Empty
+    case (Var(_, _), _) => Empty
+    case (Typ(tp1, _), Var(_, tp2)) =>
+      if (isSubType(tp2, tp1)) b else Empty
+    case (_, Var(_, _)) => Empty
+  }
+
+  /** The space of a not covered by b */
+  def minus(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) => Empty
+    case (_, Empty) => a
+    case (Typ(tp1, _), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) Empty
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1)), b)
+      else a
+    case (Typ(tp1, _), Kon(tp2, ss)) =>
+      if (isSubType(tp1, tp2)) minus(Kon(tp2, signature(tp2).map(Typ(_, false))), b)
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1)), b)
+      else a
+    case (_, Or(ss)) =>
+      ss.foldLeft(a)(minus)
+    case (Or(ss), _) =>
+      Or(ss.map(minus(_, b)))
+    case (Kon(tp1, ss), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) Empty
+      else if (simplify(a) == Empty) Empty
+      else if (isSubType(tp2, tp1) && canDecompose(tp1))
+        minus(Or(partitions(tp1)), b)
+      else a
```

Comment:
Not a bug, but potential improvement: 
```scala
trait A
trait B

abstract sealed case class Parent()
class Foo extends Parent with A
class Bar extends Parent with B
```

`Parent() - A` is known more precisely to be `Bar`